libraries/chain_kv/include/b1/chain_kv/chain_kv.hpp

```diff
@@ -110,12 +111,141 @@ inline bytes get_next_prefix(const bytes& prefix) {
    return next_prefix;
 }
 
+namespace detail {
+   typedef __uint128_t uint128_t;
+
+   template<std::size_t Size, std::size_t N>
+   void copy_and_swap_to_buffer(char* to, const char* from) {
+      char* to_ptr = to + Size * (N - 1);
+      const char* from_ptr = from;
+      for (unsigned int i = 0; i < N; ++i) {
+         memcpy(to_ptr, from_ptr, Size);
+         to_ptr -= Size;
+         from_ptr += Size;
+      }
+   }
+
+   template<typename T>
+   struct value_storage;
+
+   template<typename T>
+   struct value_storage<T*> {
+      char* as_char_ptr(T* value) const { return reinterpret_cast<char*>(value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template<typename T>
+   struct value_storage {
+      char* as_char_ptr(T& value) const { return reinterpret_cast<char*>(&value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template <typename T, std::size_t N>
+   auto append_key(bytes& dest, T value) {
+      constexpr auto type_size = value_storage<T>().size();
+      constexpr auto buf_size = type_size * N;
+      char buf[buf_size];
+      detail::copy_and_swap_to_buffer<type_size, N>(buf, value_storage<T>().as_char_ptr(value));
+      std::reverse(std::begin(buf), std::end(buf));
+      dest.insert(dest.end(), std::begin(buf), std::end(buf));
```

Comment:
Would there be any benefit in doing this operation in place in `dest`?  Expand `dest` first and then copy_and_swap and reverse on that new chunk in `dest`?

Reply:
I am presuming that the stl developers can write more performant code than me, since this isn't a specialized case for this, unless it is the case that std::begin/end and vector(dest) doesn't have an optimization.  Without knowing, I would default to the simpler approach and we can revisit when we have performance numbers. (A little hint, if know one else decides to fact check you, you can just say "know it doesn't optimize it" and I won't look it up :), I'll take your word for it)

Reply:
In an ideal world the compiler would not go through memory for the byte swapping, rending such optimizations pointless.  Unfortunately, neither gcc nor clang turns this use of std::reverse into bswap (or pshufb).